File: 04_Camp_Cleanup/04.py
import re

INPUT = "input"


def is_overlapped(assignment):
    a = set(range(assignment[0], assignment[1] + 1))
    b = set(range(assignment[2], assignment[3] + 1))
    return a.issubset(b) or b.issubset(a)


def read_input():
    with open(INPUT, "r") as file:
        assignments = file.read().split()
        assignments = [re.split(r",|-", x) for x in assignments]
        assignments = [[int(y) for y in x] for x in assignments]
    return assignments


def main():
    assignments = read_input()
    print(len(list(filter(is_overlapped, assignments))))


if __name__ == "__main__":
    assert is_overlapped([2, 4, 6, 8]) == False
    assert is_overlapped([6, 6, 4, 6]) == True

    main()


# z = x.issubset(y)

File: aoc_2022_elixir/04_Camp_Cleanup/04.exs
ExUnit.start()

defmodule AssertionTest do
  use ExUnit.Case, async: true

  test "the truth" do
    assert true
  end

  test "test is_contains" do
    assert Day04.is_contains([2, 4, 6, 8]) == false
    assert Day04.is_contains([6, 6, 4, 6]) == true
    assert Day04.is_contains([3, 7, 2, 8]) == true
    assert Day04.is_contains([2, 8, 3, 7]) == true
  end
end

defmodule Day04 do
  def read_input() do
    {:ok, file} = File.read("./input")

    file
    |> String.split()
    |> Enum.map(&String.split(&1, ["-", ","]))
    |> Enum.map(fn x -> Enum.map(x, &String.to_integer/1) end)
  end

  def is_contains([a1, a2, b1, b2]) do
    (a1 <= b1 and b2 <= a2) or (b1 <= a1 and a2 <= b2)
  end
end

assignments = Day04.read_input()

Enum.filter(assignments, &Day04.is_contains/1)
|> length()
|> IO.inspect()
